libhb/fifo.c

```diff
@@ -725,18 +725,19 @@ void hb_buffer_close( hb_buffer_t ** _b )
 #if HB_PROJECT_FEATURE_QSV
         // Reclaim QSV resources before dropping the buffer.
         // when decoding without QSV, the QSV atom will be NULL.
-        if(b->qsv_details.frame)
+        if(b->qsv_details.frame && b->qsv_details.ctx != NULL)
         {
             mfxFrameSurface1 *surface = (mfxFrameSurface1*)b->qsv_details.frame->data[3];
-            if(surface && b->qsv_details.ctx)
+            if(surface)
             {
                 if(b->qsv_details.ctx->qsv_filters_are_enabled)
                 {
+                    hb_qsv_release_surface_from_pool_by_surface_pointer(b->qsv_details.ctx->hb_vpp_qsv_frames_ctx, surface);
```

Comment:
If I'm reading the code right, the surface can only be in one of the pools (vpp or dec).  It looks like the only place b->qsv_details.frame->data[3] is assigned is in hb_qsv_copy_frame.  Could a field be added to hb_buffer_t to hold the context that owns the pool (hb_vpp_qsv_frames_ctx or hb_dec_qsv_frames_ctx) so this release only needs to be called once on the correct context?

Also, it looks like hb_qsv_release_surface_from_pool_by_surface_pointer will search the pool up to 30 times for the surface if it doesn't find it?  Not sure what that's all about, but certainly a lot of wasted CPU cycles when you know the surface won't be in the pool.

Aside from this, seems fine.

Reply:
Thanks for the comments! Done!